src/app/main.rs

```diff
@@ -190,172 +198,210 @@ where
-                    client::Config::from(ep.clone())
-                }))
-                .push(reconnect::layer());
-
-            let endpoint_stack = client_stack
-                .push(svc::stack_per_request::layer())
-                .push(normalize_uri::layer())
-                .push(orig_proto_upgrade::layer())
-                .push(tap::layer(tap_next_id.clone(), taps.clone()))
-                .push(metrics::layer::<_, ClassifyResponse>(http_metrics))
-                .push(svc::watch::layer(tls_client_config));
-
-            let dst_router_stack = endpoint_stack
-                .push(resolve::layer(Resolve::new(resolver)))
-                .push(balance::layer())
+            // TODO metrics
+            let stack = connect::Stack::new()
+                .push(control::client::Layer::new())
+                .push(control::resolve::Layer::new(dns_resolver.clone()))
+                .push(reconnect::layer().with_fixed_backoff(config.control_backoff_delay))
+                .push(proxy::timeout::layer(config.control_connect_timeout))
+                .push(svc::watch::layer(tls_client_config.clone()))
+                .push(svc::stack::phantom_data::layer())
+                .push(control::add_origin::Layer::new())
                 .push(buffer::layer())
-                .push(timeout::layer(config.bind_timeout))
-                .push(limit::layer(MAX_IN_FLIGHT))
-                .push(router::layer(Recognize::new()));
-
-            let capacity = config.outbound_router_capacity;
-            let max_idle_age = config.outbound_router_max_idle_age;
-            let router = dst_router_stack
-                .make(&router::Config::new("out", capacity, max_idle_age))
-                .expect("outbound router");
-
-            // As HTTP requests are accepted, we add some request extensions
-            // including metadata about the request's origin.
-            let server_stack = svc::stack::phantom_data::layer()
-                .push(insert_target::layer())
-                .push(timestamp_request_open::layer())
-                .bind(svc::shared::stack(router));
-
-            serve(
-                "out",
-                outbound_listener,
-                accept,
-                connect,
-                server_stack.map_err(|_| {}),
-                config.outbound_ports_disable_protocol_detection,
-                get_original_dst.clone(),
-                drain_rx.clone(),
-            )
+                .push(limit::layer(config.destination_concurrency_limit));
+
+            // Because the control client is buffered, we need to be able to
+            // spawn a task on an executor when `make` is called. This is done
+            // lazily so that a default executor is available to spawn the
+            // background buffering task.
+            future::lazy(move || match control_config {
+                None => future::ok(None),
+                Some(config) => match stack.make(&config) {
```

Comment:
Nit: Since this now returns `Result<Option<...>, ...>` rather than `Option<...>`, I think this match can be reduced to something like 
```suggestion
                Some(config) => stack
                    .make(&config)
                    .map(Some)
                    .map_err(|e| { error!("failed to build controller: {}", e); })
```

Reply:
(Also, IIRC, it's not strictly necessary to use `future::ok`/`future::err` here; `future::lazy` takes a closure returning `IntoFuture`, which `Ok` and `Err` implement)